lib/CFRPageActions.jsm

```diff
@@ -288,12 +273,82 @@ class PageAction {
+  async _renderPopup(message, browser) {
+    const {id, content} = message;
+
+    const headerLabel = this.window.document.getElementById("cfr-notification-header-label");
+    const headerLink = this.window.document.getElementById("cfr-notification-header-link");
+    const headerImage = this.window.document.getElementById("cfr-notification-header-image");
+    const footerText = this.window.document.getElementById("cfr-notification-footer-text");
+    const footerLink = this.window.document.getElementById("cfr-notification-footer-learn-more-link");
+
+    headerLabel.value = await this.getStrings(content.heading_text);
+    headerLink.setAttribute("href", SUMO_BASE_URL + content.info_icon.sumo_path);
+    headerLink.setAttribute(this.window.RTL_UI ? "left" : "right", 0);
+    headerImage.setAttribute("tooltiptext", await this.getStrings(content.info_icon.label, "tooltiptext"));
+    headerLink.onclick = () => this._sendTelemetry({message_id: id, bucket_id: content.bucket_id, event: "RATIONALE"});
 
+    let primaryActionCallback;
+    let options = {};
+    let panelTitle;
     const {primary, secondary} = content.buttons;
+
+    footerText.textContent = await this.getStrings(content.text);
+
+    if (content.addon) {
+      await this._setAddonAuthorAndRating(this.window.document, content);
+      panelTitle = await this.getStrings(content.addon.title);
+
+      footerLink.value = await this.getStrings({string_id: "cfr-doorhanger-extension-learn-more-link"});
+      footerLink.setAttribute("href", content.addon.amo_url);
+      footerLink.onclick = () => this._sendTelemetry({message_id: id, bucket_id: content.bucket_id, event: "LEARN_MORE"});
+
+      primaryActionCallback = async () => {
+        primary.action.data.url = await CFRPageActions._fetchLatestAddonVersion(content.addon.id); // eslint-disable-line no-use-before-define
+        this._blockMessage(id);
+        this.dispatchUserAction(primary.action);
+        this.hideAddressBarNotifier();
+        this._sendTelemetry({message_id: id, bucket_id: content.bucket_id, event: "INSTALL"});
+        RecommendationMap.delete(browser);
+      };
+
+      options = {
+        popupIconURL: content.addon.icon,
+        hideClose: true,
+        eventCallback: this._popupStateChange,
+      };
+
+    } else {
+      const stepsContainerId = "cfr-notification-feature-steps";
+      primaryActionCallback = () => {
+        this._blockMessage(id);
+        this.dispatchUserAction(primary.action);
+        this.hideAddressBarNotifier();
+        this._sendTelemetry({message_id: id, bucket_id: content.bucket_id, event: "PIN"});
+        RecommendationMap.delete(browser);
+      };
+      panelTitle = await this.getStrings(content.heading_text);
+      let stepsContainer = this.window.document.getElementById(stepsContainerId);
+
+      // Container for bullet point list
+      if (!stepsContainer) {
```

Comment:
I think you probably intended to only have `stepsContainer = this.window.document.createElement("vbox")` under this if statement? Right now the behaviour here will be that the text never gets changed if a second CFR shows up